draft-ietf-httpbis-resumable-upload.md

```diff
@@ -423,6 +423,16 @@ The `Upload-Complete` header field MUST only be used if support by the resource
 
 The `301 (Moved Permanently)` and `302 (Found)` status codes MUST NOT be used in offset retrieval ({{offset-retrieving}}) and upload cancellation ({{upload-cancellation}}) responses. For other responses, the upload resource MAY return a `308 (Permanent Redirect)` status code and clients SHOULD use new permanent URI for subsequent requests. If the client receives a `307 (Temporary Redirect)` response to an offset retrieval ({{offset-retrieving}}) request, it MAY apply the redirection directly in an immediate subsequent upload append ({{upload-appending}}).
 
+# Integrity Digests
+
+The integrity of an entire upload or individual upload requests can be verifying using digests from {{!DIGEST-FIELDS=RFC9530}}.
+
+If the client knows the integrity digest of the entire data before creating an upload resource, it MAY include the `Repr-Digest` header field when creating an upload ({{upload-creation}}). Once the upload is completed, the server can compute the integrity digest of the received upload representation and compare it to the provided digest. If the digests don't match the server SHOULD consider the transfer failed and not process the uploaded data further. This way, the integrity of the entire uploaded data can be protected.
+
+If the client knows the integrity digest of the content from an upload creation ({{upload-creation}}) or upload appending ({{upload-appending}}) request, it MAY include the `Content-Digest` header field in the request. Once the content has been received, the server can compute the integrity digest of the received content and compare it to the provided digest. If the digests don't match the server SHOULD consider the transfer failed and not append the content to the upload resource. This way, the integrity of an individual request can be protected.
+
+If the client does not know the integrity digest of the content from an upload creation ({{upload-creation}}) or upload appending ({{upload-appending}}) request, it MAY request the integrity digest from the server by including the `Want-Content-Digest` header field. While transferring the content, the client can compute the integrity digest of the transferred content and compare it to the digest provided by the server in the response. If the digests don't match the client SHOULD consider the transfer failed and cancel the upload ({{upload-cancellation}}).
+
```

Comment:
Is there case for the situation where the server doesn't want to verify repr-digest itself but can calculate one and make it available for the client to obtain? For example, over a series of upload appends it combines ranges (content) into a representation. Then once upload is complete, the server can calculate the complete repr-digest and return that, allowing the client to perform a comparison of the local and remote views.

Reply:
> Then once upload is complete, the server can calculate the complete repr-digest and return that, allowing the client to perform a comparison of the local and remote views.

Yes, I could image that very well. A client may calculate the checksum while uploading the data and thus cannot include it in the request header fields. In addition, comparing the complete repr-digest has the benefit that it verifies that the messages from the upload append requests were properly combined into a final representation, whereas comparing the repr-digest of a single upload append request only verifies the integrity of the message. If there is a bug on the server-side combining the upload appends, then this would be caught be the comparison of the complete repr-digest.

Overall, I could image that Repr-Digest is more helpful than Content-Digest for resumable uploads. If an upload append (or an upload creation) uses Content-Digest to secure the integrity of the message and the request transmission is interrupted in the middle, the server likely received just a partial message content and its digest won't match the digest of the entire message content that the client included in Content-Digest. A server may then reject the partial message due to the failure in validating the integrity, forcing the client to re-transmit the data.

The use of Repr-Digest helps in this case because the upload can interrupted and resumed multiple times while also accepting partial messages. Only once the upload is finished, the Repr-Digest of the entire representation is compared (either by the client or the server).

Reply:
On re-reading, I don't think the use of `Want-Content-Digest` as described is standard. A client's "want-content-digest" would request that the server add `Content-Digest` over the response message content. Unless the response contains the bytes that were uploaded, the content-digest won't meet the expectations in your proposed text. Requiring a server to echo the actual upload chunk doesn't seem helpful to me. Can we punt on this suggestion and consider it in a follow up?